build_tools/python/e2e_test_framework/serialization.py

```diff
@@ -0,0 +1,249 @@
+                 keyed_obj: bool = False,
+                 type_key: Optional[str] = None,
+                 id_field: str = "id"):
+  """Decorator to make a dataclass serializable.
+  
+  Args:
+    keyed_obj: is the class a keyed object, which is unique per id and will only
+      have one copy in the serialization per id.
+    type_key: string defining the object type, must be set if keyed_obj is True.
+    id_field: field name of the id field of a keyed object.
+
+  Example:
+    @serializable
+    @dataclass
+    class A(object):
+      ...
+
+    @serialzable(keyed_obj=True, type_key="obj_b")
+    @dataclass
+    class B(object):
+      id: str
+  """
+
+  if keyed_obj and type_key is None:
+    raise ValueError("type_key must be set if keyed_by_id is true.")
+  if type_key is not None and ":" in type_key:
+    raise ValueError("':' is the reserved character in type_key.")
+
+  def wrap(cls):
+    if not dataclasses.is_dataclass(cls):
+      raise ValueError(f"{cls} is not a dataclass.")
+
+    fields = dataclasses.fields(cls)
+
+    def serialize(self, keyed_obj_map: OrderedDict[str, Any]):
+      if not keyed_obj:
+        return _fields_to_dict(self, fields, keyed_obj_map)
+
+      obj_id = getattr(self, id_field)
+      # type_key has been checked to be not None above.
+      assert type_key is not None
+      obj_key = f"{type_key}:{obj_id}"
+      if obj_key in keyed_obj_map:
+        # If the value in the map is None, it means we have visited this object
+        # before but not yet finished serializing it. This will only happen if
+        # there is a circular reference.
+        if keyed_obj_map[obj_key] is None:
+          raise ValueError(f"Circular reference is not supported: {obj_key}.")
+        return obj_id
+
+      # Populate the keyed_obj_map with None first to detect circular reference.
+      keyed_obj_map[obj_key] = None
+      obj_dict = _fields_to_dict(self, fields, keyed_obj_map)
+      keyed_obj_map[obj_key] = obj_dict
+      return obj_id
+
+    def deserialize(data, keyed_obj_map: Dict[str, Any], obj_cache: Dict[str,
+                                                                         Any]):
+      if not keyed_obj:
+        field_value_map = _dict_to_fields(data, fields, keyed_obj_map,
+                                          obj_cache)
+        return cls(**field_value_map)
+
+      obj_id = data
+      # type_key has been checked to be not None above.
+      assert type_key is not None
+      obj_key = f"{type_key}:{obj_id}"
+      if obj_key in obj_cache:
+        return obj_cache[obj_key]
+
+      field_value_map = _dict_to_fields(keyed_obj_map[obj_key], fields,
+                                        keyed_obj_map, obj_cache)
+      derialized_obj = cls(**field_value_map)
+      obj_cache[obj_key] = derialized_obj
+      return derialized_obj
+
+    setattr(cls, SERIALIZE_FUNC_NAME, serialize)
+    setattr(cls, DESERIALIZE_FUNC_NAME, deserialize)
+    return cls
+
+  if cls is None:
```

Comment:
? That seems like an odd condition

Reply:
This is the trick to make the uses of both `@serializable` and `@serializable(...)` possible. Documented.

Reply:
Aaaah I get it